### PR TITLE
fourth stage ready

### PR DESCRIPTION
**Pull request by agavriluta**
image

### CODE REVIEW

File: MediumProject/pom.xml [hunk 1]
```diff
@@ -19,8 +19,8 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <configuration>
-                    <source>11</source>
-                    <target>11</target>
+                    <source>16</source>
```
Comment (ppanasencoGD) [inline MediumProject/pom.xml:23]: please use single version of java and not changing it for each stage

File: MediumProject/src/main/java/Main.java [hunk 2]
```diff
@@ -1,7 +1,22 @@
...
+import java.io.File;
+import java.util.Objects;
 
 public class Main {
     public static void main(String[] args) {
-        Searching.search();
+       startStage4(args);
+    }
+
+    private static void startStage4(String[] args) {
```
Comment (ppanasencoGD) [inline MediumProject/src/main/java/Main.java]: rename

File: MediumProject/src/main/java/stage4/SearchingEngine.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    public static void searchPerson(String searchingWord){
+        List<Person> foundUsers = searchUsers( searchingWord);
+        printSearchResult(foundUsers);
+    }
+
+    private static List<Person> searchUsers( String searchingWord) {
+        return generalListOfUsers.stream().findAny()
+                .stream()
+                .filter(p -> p.getFirstName().toLowerCase().contains(searchingWord) ||
```
Comment (ppanasencoGD) [inline MediumProject/src/main/java/stage4/SearchingEngine.java]: will it work if you add toString() or custom getPersonDetails() method which will concat all 3 attributes

                .filter(p -> p.getFirstName().toLowerCase().contains(searchingWord) ||
                        p.getLastName().toLowerCase().contains(searchingWord) ||
                        p.getEmail().toLowerCase().contains(searchingWord))

=>>>>>

                 .filter(p -> p.getPersonDetails().toLowerCase().contains(searchingWord) )

in Person class 

 public String getPersonDetails(){
return this.firstName+this.lastName+.....
}

**Review by ppanasencoGD — APPROVED** [review]

File: MediumProject/src/main/java/stage4/SearchingEngine.java (hunk 3 above)
Comment (ppanasencoGD) [reply MediumProject/src/main/java/stage4/SearchingEngine.java]: btw you can use toString() -


    @Override
    public String toString() {
        if (email.isBlank()) return firstName + " " + lastName;
        else return firstName + " " + lastName + " " + email;
    }

File: MediumProject/src/main/java/stage4/SearchingEngine.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+public class SearchingEngine {
+    private static final List<Person> generalListOfUsers = List.copyOf(UserDetails.getPeople());
+
+    public static void searchPerson(String searchingWord){
+        List<Person> foundUsers = searchUsers( searchingWord);
+        printSearchResult(foundUsers);
+    }
+
+    private static List<Person> searchUsers( String searchingWord) {
+        return generalListOfUsers.stream().findAny()
```
Comment (ppanasencoGD) [inline MediumProject/src/main/java/stage4/SearchingEngine.java:12]: why findAny()????

File: MediumProject/pom.xml (hunk 1 above)
Comment (agavriluta) [reply MediumProject/pom.xml:23]: I always have such errors 
image

File: MediumProject/src/main/java/stage4/SearchingEngine.java (hunk 4 above)
Comment (agavriluta) [reply MediumProject/src/main/java/stage4/SearchingEngine.java:12]: changed to findFirst.

File: MediumProject/src/main/java/stage4/SearchingEngine.java (hunk 3 above)
Comment (agavriluta) [reply MediumProject/src/main/java/stage4/SearchingEngine.java]: done

File: MediumProject/src/main/java/Main.java (hunk 2 above)
Comment (agavriluta) [reply MediumProject/src/main/java/Main.java]: named startSearchingStage4
is it ok ?

**Pull request merged**